third_party/nvidia/lib/Dialect/NVWS/Transforms/LowerAref.cpp

```diff
@@ -0,0 +1,725 @@
+      } else {
+        // persistent
+        // Get should be inside of a for loop
+        auto forOp = dyn_cast<scf::ForOp>(mmaGetOp->getBlock()->getParentOp());
+        if (!forOp)
+          return failure();
+        // TODO: relax the assumption about which position the loop-carried
+        // iteration count is in
+        idx = forOp.getOperand(1);
+        rewriter.setInsertionPoint(&mmaGetOp.getRegion().front().front());
+      }
+      waitOnMbar(rewriter, loc, getArefValue.emptyMbars,
+                 rewriter.create<arith::ConstantIntOp>(loc, 0, 32), idx,
+                 getArefValue.depth);
+    }
+
+    // Since we are signalling the inputs to the mma are done with the mma
+    // barrier, we don't need to add an arrival when lowering the get.
+    graph.nodes[getOp].containsAsync = true;
+
+    return success();
+  }
+};
+
+class ParallelizeWarpGroupDot : public OpRewritePattern<WarpGroupDotOp> {
+  ArefUseGraph &graph;
+
+public:
+  using OpRewritePattern::OpRewritePattern;
+
+  ParallelizeWarpGroupDot(mlir::MLIRContext *context, ArefUseGraph &graph)
+      : OpRewritePattern<WarpGroupDotOp>(context), graph(graph) {}
+
+  LogicalResult matchAndRewrite(WarpGroupDotOp op,
+                                PatternRewriter &rewriter) const override {
+    auto ctx = op.getContext();
+    auto loc = op.getLoc();
+    auto getOp = dyn_cast<ArefGetOp>(op->getBlock()->getParentOp());
+    if (!getOp)
+      return failure();
+
+    auto kLoop = dyn_cast<scf::ForOp>(getOp->getBlock()->getParentOp());
+    if (!kLoop)
+      return failure();
+
+    auto putOp = dyn_cast<ArefPutOp>(kLoop->getBlock()->getParentOp());
+    if (!putOp)
+      return failure();
+
+    auto putAref = putOp.getOperand();
+    auto getAref = getOp.getOperand();
+    if (!graph.arefs.contains(putAref) || !graph.arefs.contains(getAref))
+      return failure();
+
+    auto getArefValue = graph.arefs[getAref];
+    auto putArefValue = graph.arefs[putAref];
+
+    if (putAref.getType().getBaseType().size() != 1)
+      return failure(
+          "TODO: support putting into more than one accumulator at a time");
+
+    SmallVector<Operation *> users(op->user_begin(), op->user_end());
+    if (users.size() != 1)
+      return failure();
+    auto wait = dyn_cast<WarpGroupDotWaitOp>(users[0]);
+    if (!wait)
+      return failure();
+
+    if (wait.getPendings() == (getArefValue.depth - 1))
+      return failure(); // we've already processed this
+
+    wait.setPendings(getArefValue.depth - 1);
+    rewriter.setInsertionPointAfter(wait);
+
+    // This assumes inputs to an mma are both from a single aref get. Need to
+    // relax that to allow nested aref gets for attention
+    Value getIdx = getArefValue.depth > 1
+                       ? getOp.getIndexes()[0]
+                       : rewriter.create<arith::ConstantIntOp>(loc, 0, 32);
+    auto depthVal =
+        rewriter.create<arith::ConstantIntOp>(loc, getArefValue.depth - 1, 32);
```

Comment:
Is this depth the multiplicity of SMEM buffers or the MMA pipeline depth? For hopper it needs to be the latter.